---
title: "Analysis \"Template\""
output: 
  html_notebook:
    author: "David McGaughey"
    date: `r Sys.Date()`
    theme: flatly
    toc: true
    toc_float: true
    code_folding: hide
---
```{r, message=FALSE, message=FALSE, warning=FALSE}
library(data.table)
library(tidyverse)
library(biomaRt)
library(tximport)
library(DESeq2)
library(stringr)
library(ggrepel)
library(apeglm)
library(clusterProfiler)
# Parallel
library(BiocParallel)
register(MulticoreParam(6))
working_dir <- '~/data/project/salmon_quant/'
files <- list.files(path=working_dir,recursive=TRUE,pattern='quant.sf', full.names = TRUE)
```

# Extract gene annotation
```{r, message=FALSE, message=FALSE, warning=FALSE}
anno <- fread(files[1]) 
anno$Gene <- sapply(anno$Name,function(x) strsplit(x,'\\|')[[1]][6])
anno_tximport <- anno %>% 
  dplyr::select(target_id = Name, Gene)
```

# Import quant at gene level
```{r, message=FALSE, message=FALSE, warning=FALSE}
txi <- tximport(files, type = "salmon", tx2gene = anno_tximport)
txi.deseq2 <- data.frame(txi$counts)
```

# Metadata
```{r}
meta <- gsub(pattern = '\\/Users\\/mcgaugheyd\\/data\\/project\\/salmon_quant\\/\\/','',files) %>% gsub('\\/quant.sf','',.) %>% enframe()
meta$fastq_path <- files
meta <- meta %>% separate(value, into = c('line','treatment','biological_replicate'), sep = '_', remove = FALSE)
```

# Deseq2 time
Treatment compared with line as a covariate

VST norm for PCA
```{r, message=FALSE, message=FALSE, warning=FALSE}
dds <- DESeqDataSetFromTximport(txi, meta, ~treatment + line)

DESeq2Table <- DESeq(dds, parallel = T)
# rld <- rlogTransformation(DESeq2Table, blind=TRUE)
vst <- varianceStabilizingTransformation(DESeq2Table)
colnames(vst) <- meta$value
```

# PCA
```{r}
#pca 
library(matrixStats)
ntop = 1000
Pvars <- rowVars(assay(vst))
select <- order(Pvars, decreasing = TRUE)[seq_len(min(ntop, 
                                                      length(Pvars)))]
PCA <- prcomp(t(assay(vst)[select, ]), scale = F)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
dataGG = data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2], 
                    PC3 = PCA$x[,3], PC4 = PCA$x[,4], 
                    line = colData(vst)$line,
                    treatment=colData(vst)$treatment)

ggplot(dataGG, aes(PC1, PC2, color=line, shape=treatment)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  cowplot::theme_cowplot()

ggplot(dataGG, aes(PC3, PC4, color=line, shape=treatment)) +
  geom_point(size=3) +
  xlab(paste0("PC3: ",percentVar[3],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  cowplot::theme_cowplot()
```

## PC Loadings
PC2 appears to separate BLAH against the rest

Here are the genes which contribute most to this PC:
```{r}
PCA$rotation %>% as_tibble(rownames = 'Gene') %>% arrange(-abs(PC2)) %>% dplyr::select(Gene,PC2) %>% head(50) %>% DT::datatable()
```

## Visualize Loadings
```{r}
pcFirst <- 'PC1'
pcSecond <- 'PC2'
rotations <- c(pcFirst, pcSecond)
rotation_multipler <- 100 # arbitrary value to make the loading vectors about the right size on the plot
# rotations for pc1 and pc2
top_rotations <- 
  c(PCA$rotation[,str_extract(pcFirst, '\\d+') %>% as.integer()] %>% sort() %>% head(3) %>% names(),
    PCA$rotation[,str_extract(pcFirst, '\\d+') %>% as.integer()] %>% sort() %>% tail(3) %>% names(),
    PCA$rotation[,str_extract(pcSecond, '\\d+') %>% as.integer()] %>% sort() %>% head(3) %>% names(),
    PCA$rotation[,str_extract(pcSecond, '\\d+') %>% as.integer()] %>% sort() %>% tail(3) %>% names()) %>% 
  unique()

dataGG %>%
  as_tibble() %>% 
  ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
  geom_point(size=3, aes(color=line, shape = treatment)) +
  geom_segment(data = PCA$rotation[top_rotations,rotations] %>% data.frame(), aes(x=0,y=0, xend = .data[[pcFirst]]*rotation_multipler, yend = .data[[pcSecond]]*rotation_multipler)) +
  ggrepel::geom_label_repel(data = PCA$rotation[top_rotations,rotations] %>% 
                              as_tibble(rownames = 'Gene') %>% 
                              mutate(Gene = gsub(' \\(.*','',Gene)), 
                            aes(x=.data[[pcFirst]]*rotation_multipler, y = .data[[pcSecond]] * rotation_multipler, label = Gene)) +
  xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
  ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) + 
  cowplot::theme_cowplot() 
```

# Heatmap
Run clustering on sample <-> sample distances

```{r, fig.height=4, fig.width=5}
library(ComplexHeatmap)
sampleDists <- dist(t(assay(vst)[select, ]))

sampleDistMatrix <- as.matrix(sampleDists) %>% scale(., center= FALSE)
colnames(sampleDistMatrix) <- colData(dds)$Sample
row.names(sampleDistMatrix) <- colData(dds)$Sample

ha_column = HeatmapAnnotation(df = data.frame(Cells = meta$Cells %>% factor(),
                                                  Treatment = meta$Treatment %>% factor()),
                                  col = list(Cells = c("Line A" = viridis::viridis(20)[6], 
                                                       "Line B" = viridis::viridis(20)[14]),
                                             Treatment = c("CTRL" = viridis::magma(20)[3], 
                                                           "BLEACH" = viridis::magma(20)[7], 
                                                           "UNICORN HORNS" = viridis::magma(20)[11],
                                                           "POWERFUL FLASHLIGHT" =  viridis::magma(20)[15])))


Heatmap(sampleDistMatrix, 
        top_annotation = ha_column,name = 'Distance',
        col = circlize::colorRamp2(seq(0,1.4,0.1), 
                                   colors = viridis::viridis(n = 15)))
```

# Diff Testing

Run Diff Testing with a contrast to remove one of the treatments
``` {r, echo=FALSE,warning=FALSE,message=FALSE}
#########
# DE time
#########

design(dds) <- formula(~ type + condition)
dds <- DESeq(dds)
res <- results(dds, contrast = c('treatment','A','B'))

resLFC <- lfcShrink(dds, coef = "treatment_A_vs_B", type = 'apeglm')
plotMA(resLFC)
resLFC$pvalue %>% hist() %>% plot()
```

## Results
Positive is genes that are more expressed in A vs B
```{r}
resLFC %>%  as_tibble(rownames = 'Gene') %>% arrange(padj) %>% head(2000) %>% DT::datatable()
```
## Volcano Plot
On the right is genes that are more expressed in A vs B
```{r}
volcano_maker <- function(df, title="Volcano Plot", pvalue='pvalue', padj='padj', logFC='log2FoldChange', gene_list = ''){
  df <- df[!is.na(df$pvalue),]
  print(dim(df))
  df$Class <- 'Not Significant'
  df$Class[df[,'padj'] < 0.01] <- "FDR < 0.01"
  df$GeneT <- df$Gene
  df$Gene[!df$Gene %in% gene_list] <- ''
  
  plot <- ggplot(data=df,aes(label=Gene, x = log2FoldChange, y = -log10(pvalue))) + 
    geom_point(aes(colour=Class)) +
    scale_colour_manual(values=c("darkred","grey")) + 
    cowplot::theme_cowplot() +
    geom_label_repel(data=bind_rows(subset(df, padj<0.01 & baseMean > 500) %>% arrange((log2FoldChange)) %>% head(10), 
                                    subset(df, padj<0.01 & baseMean > 500) %>% arrange((log2FoldChange)) %>% tail(10)), 
                     aes(label=GeneT)) +
    xlab('logFC') + ylab('-log10(p value)') +
    ggtitle(title) + cowplot::theme_cowplot() +
    geom_vline(aes(xintercept=-1),linetype="dotted") +
    geom_vline(aes(xintercept=1),linetype="dotted") 
  plot
}

volcano_maker(df = resLFC %>% as_tibble(rownames = 'Gene')) + coord_cartesian(xlim = c(-5,5))
```

## Dot Plots of Select Genes
```{r, fig.width=5, fig.height=3}
gene_TPM <- txi$abundance
colnames(gene_TPM) <- meta$value
plot_genes <- bind_rows(subset(resLFC %>% as_tibble(rownames = 'Gene'), padj<0.01 & baseMean > 500 & log2FoldChange > 1) %>% arrange((padj)) %>% head(10), 
                        subset(resLFC %>% as_tibble(rownames = 'Gene'), padj<0.01 & baseMean > 500 & log2FoldChange < -1) %>% arrange((padj)) %>% head(10)) %>% pull(Gene)
gene_TPM %>% as_tibble(rownames = 'Gene') %>% 
  pivot_longer(-Gene) %>% 
  left_join(meta, by = c('name' = 'value')) %>% 
  filter(Gene %in% plot_genes) %>% 
  ggplot(aes(x=Condition, y = log2(value), color = name)) + 
  geom_boxplot(color = 'black') +
  geom_point() + 
  facet_wrap(~Gene) +
  cowplot::theme_cowplot() +
  coord_flip()
```

# Enrichment Analysis (GO and Pathway)

## GO Enrichment 
```{r}
library(clusterProfiler)
library(enrichplot)


diff_genes <- resLFC %>% as_tibble(rownames = 'Gene') %>% filter(padj < 0.01, abs(log2FoldChange) > 1) 
eg_diff_genes <- bitr(diff_genes$Gene, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
eg_diff_genes <- diff_genes %>% left_join(., eg_diff_genes, by = c('Gene' = 'SYMBOL'))
# only keep genes that have a mean of >= 10 reads across all sample
expressed <- txi$counts[txi$counts %>% rowMeans() > 10,] %>% row.names()
eg_universe = bitr(resLFC %>% as_tibble(rownames = 'Gene') %>% filter(Gene %in% expressed) %>% pull(Gene), fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
ego <- enrichGO(gene          = eg_diff_genes$ENTREZID,
                universe      = eg_universe$ENTREZID,
                OrgDb         = org.Hs.eg.db,
                ont           = "all",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable      = TRUE)
```

```{r, fig.width=3, fig.height=5}
p1 <- dotplot(ego, showCategory=30) + ggtitle("Dotplot for GO")
p1
```

### Table
```{r}
ego <- setReadable(ego, OrgDb = org.Hs.eg.db, keyType = 'ENTREZID')
ego@result %>% DT::datatable()
```

### "CNET" Plot
Relationships between related GO terms with shared genes
```{r, fig.width=6, fig.height=4}
geneList <- eg_diff_genes$log2FoldChange
names(geneList) <- eg_diff_genes$Gene
cnet <- cnetplot(ego, foldChange = geneList) + scale_color_viridis_c(name = 'log2(FoldChange)')
cnet
```

### Enriched GO Term Relationships
Shows the "sets" of GO terms to increase understanding of what the diff gene sets are doing. There are many related GO terms.
```{r, fig.width=8, fig.height=8}
ego_pair <- pairwise_termsim(ego) 
go_term_relationships <- emapplot(ego_pair,layout="kk") #+ ggtitle("Enriched GO Term Relationships")
go_term_relationships
```
## Wikipathways
```{r}
# system("wget https://wikipathways-data.wmcloud.org/current/gmt/wikipathways-20230210-gmt-Homo_sapiens.gmt")
wp2gene <- read.gmt('wikipathways-20230210-gmt-Homo_sapiens.gmt')
wp2gene <- wp2gene %>% tidyr::separate(term, c("name","version","wpid","org"), "%")
wpid2gene <- wp2gene %>% dplyr::select(wpid, gene) #TERM2GENE
wpid2name <- wp2gene %>% dplyr::select(wpid, name) #TERM2NAME
ewp <- enricher(eg_diff_genes$ENTREZID,
                TERM2GENE = wpid2gene,
                TERM2NAME = wpid2name,
                pvalueCutoff = 0.1)
ewp_plot <- dotplot(ewp, showCategory=30) + ggtitle("Dotplot for WikiPathways")
```

```{r, fig.width=4, fig.height=3}
ewp_plot
```
### Table
```{r}
ewp <- setReadable(ewp, OrgDb = org.Hs.eg.db, keyType = 'ENTREZID')
ewp@result %>% DT::datatable()
```

## GSEA
```{r}
all_genes <- bitr(resLFC %>% as_tibble(rownames = 'Gene') %>% filter(Gene %in% expressed) %>% pull(Gene), fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
all_genes <- all_genes %>% left_join(resLFC %>% as_tibble(rownames = 'SYMBOL'), by = c('SYMBOL'))
logFC <- all_genes$log2FoldChange
names(logFC) <- all_genes$ENTREZID
logFC <- na.omit(logFC)
logFC = sort(logFC, decreasing = TRUE)
gse <- gseGO(geneList=logFC,
             ont ="ALL",
             keyType = "ENTREZID",
             pvalueCutoff = 0.05,
             OrgDb = org.Hs.eg.db,
             pAdjustMethod = "BH")
```

```{r, fig.width=4, fig.height=5}
dotplot(gse, showCategory=10, split=".sign") + facet_grid(.~.sign)
```

### Table
```{r}
gse <- setReadable(gse, OrgDb = org.Hs.eg.db)
gse@result %>% as_tibble() %>% arrange(p.adjust) %>% filter(p.adjust < 0.01) %>% DT::datatable()
```


### Heatmaps

#### Extracellular Matrix (GO:0031012)
```{r}
relative <- log2(gene_TPM + 1) # add a pseudocount of 1 as log2(0) is infinity
row_means <- rowMeans(relative)
relative <- relative - row_means
#relative_g <-  relative

gene_set<- (gse@result %>% filter(ID == 'GO:0031012') %>% pull(core_enrichment) %>% str_split(., '\\/'))[[1]]
```

## KEGG Pathway Enrichment
```{r, fig.width=3}
kk <- enrichKEGG(gene         = eg_diff_genes$ENTREZID, 
                 universe = eg_universe$ENTREZID,
                 organism     = 'hsa',
                 pvalueCutoff = 0.1)
dotplot(kk) + ggtitle("KEGG Pathway Enrichment") 
```

### Table
```{r}
kk <- setReadable(kk, OrgDb = org.Hs.eg.db, keyType = 'ENTREZID')
kk@result %>% as_tibble() %>% arrange(p.adjust) %>% filter(p.adjust < 0.01) %>% DT::datatable()
```

### Focal Adhesion (hsa04510)
```{r}
gene_set<- (kk@result %>% filter(ID == 'hsa04510') %>% pull(geneID) %>% str_split(., '\\/'))[[1]]
```

```{r, fig.width=3, fig.height=3}
hm_matrix <- relative[gene_set,]

Heatmap(hm_matrix, 
        cluster_columns = FALSE,
        cluster_rows = TRUE, 
        name = 'log2(Fold\nChange)', 
        top_annotation = ha_column)
```


# Session Info
```{r}
devtools::session_info()
```
